src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/transformers/ThreadTransformer.kt

```
@@ -57,6 +66,40 @@ internal class ThreadTransformer(
         visitInsn(opcode)
     }
 
+    override fun visitMaxs(maxStack: Int, maxLocals: Int) = adapter.run {
+        if (isThreadRunMethod(methodName, desc)) {
+            visitLabel(runMethodTryBlockEnd)
+            visitLabel(runMethodCatchBlock)
+
+            // STACK: exception
+            dup()
+            invokeStatic(Injections::onThreadException)
+            // STACK: exception, isSuppressed
+
+            // Notify that the thread has finished.
+            // TODO: currently does not work, because `ManagedStrategy::onThreadFinish`
+            //   assumes the thread finish injection is called under normal managed execution,
+            //   i.e., in non-aborted state
+            // invokeStatic(Injections::afterThreadFinish)
+
+            // Suppress exception if necessary.
+            val suppressLabel = newLabel()
+            // STACK: exception, isSuppressed
+            ifZCmp(NE, suppressLabel)
+
+            // Re-throw exception
+            // STACK: exception
+            visitInsn(ATHROW)
```

Let's throw the exception in `onThreadException`